### PR TITLE
RavenDB-20137 - Invalid query results when using paging / different results on different nodes

### PR DESCRIPTION
**Pull request by grisha-kotler**
Since we are doing a binary search, we must preserve the order of the terms.
Since we have a term that we don't want to allocate and in order to preserve the order, we are going to use the previous term.

### CODE REVIEW

File: src/Lucene.Net/Util/UnmanagedStringArray.cs [hunk 1]
```diff
@@ -215,6 +215,8 @@ public void Add(Span<char> str)
 
         public void AddEmpty()
         {
+            // since we are doing a binary search, we must keep the order of the terms
+            _strings[_index].Start = _index > 0 ? _strings[_index - 1].Start : null;
```
Comment (arekpalinski) [inline src/Lucene.Net/Util/UnmanagedStringArray.cs]: For `_index == 0` we'll always have `null`. This means that in binary search this item will be always considered the lowest, won't it?. I have some doubts whether that is fine. We won't preserve the original order. Maybe for element at `0` position we need to _always_ add the real term value

CC @maciejaszyk

**Comment by arekpalinski** [comment]
For reference - this is fixing regression introduced in https://github.com/ravendb/lucenenet/pull/47

File: src/Lucene.Net/Util/UnmanagedStringArray.cs (hunk 1 above)
Comment (arekpalinski) [reply src/Lucene.Net/Util/UnmanagedStringArray.cs]: Also current fix won't work for element at position `1` since at `_index - 1` will be `null`

Comment (maciejaszyk) [reply src/Lucene.Net/Util/UnmanagedStringArray.cs]: I would add the original term at that place. Should not decrease performance but give some certainty about the correctness of the order, isn't it?

Comment (grisha-kotler) [reply src/Lucene.Net/Util/UnmanagedStringArray.cs]: You are correct. Changed.

**Comment by arekpalinski** [comment]
Merging the PR. Test will be added in separate PR

**Pull request merged**